samples/msal-node-samples/client-credentials-distributed-cache/src/AuthProvider.ts

```diff
@@ -0,0 +1,148 @@
+        const cca = new ConfidentialClientApplication({
+            ...this.msalConfig,
+            cache: {
+                cachePlugin: new CustomCachePlugin(
+                    this.cacheClientWrapper,
+                    `${this.msalConfig.auth.clientId}.${this.msalConfig.auth.authority!.split("/").pop()!}` // partitionKey <clientId>.<tenantId>
+                )
+            }
+        });
+
+        let tokenResponse = null;
+
+        try {
+            performance.mark("acquireTokenByClientCredential-start");
+            await cca.getTokenCache().getAllAccounts(); // required for triggering cache read
+            tokenResponse = await cca.acquireTokenByClientCredential(tokenRequest);
+            performance.mark("acquireTokenByClientCredential-end");
+            performance.measure(
+                tokenResponse?.fromCache ? "acquireTokenByClientCredential-fromCache" : "acquireTokenByClientCredential-fromNetwork",
+                "acquireTokenByClientCredential-start",
+                "acquireTokenByClientCredential-end"
+            );
+        } catch (error) {
+            throw error;
+        }
+
+        return tokenResponse;
+    }
+
+    private static async getMetadata(msalConfig: Configuration, cacheClient: RedisClientType): Promise<Configuration> {
+        const msalConfigWithMetadata = msalConfig;
+        const clientId = msalConfigWithMetadata.auth.clientId;
+        const tenantId = msalConfigWithMetadata.auth.authority!.split("/").pop()!;
+
+        try {
+            let [cloudDiscoveryMetadata, authorityMetadata] = await Promise.all([
+                cacheClient.get(`${clientId}.${tenantId}.discovery-metadata`),
+                cacheClient.get(`${clientId}.${tenantId}.authority-metadata`)
+            ]);
+
+            if (!cloudDiscoveryMetadata || !authorityMetadata) {
+                [cloudDiscoveryMetadata, authorityMetadata] = await Promise.all([
+                    AuthProvider.fetchCloudDiscoveryMetadata(tenantId),
+                    AuthProvider.fetchAuthorityMetadata(tenantId)
+                ]);
+
+                if (cloudDiscoveryMetadata && authorityMetadata) {
+                    await cacheClient.set(`${clientId}.${tenantId}.discovery-metadata`, JSON.stringify(cloudDiscoveryMetadata));
+                    await cacheClient.set(`${clientId}.${tenantId}.authority-metadata`, JSON.stringify(authorityMetadata));
+                }
+            }
+
+            msalConfigWithMetadata.auth.cloudDiscoveryMetadata = typeof cloudDiscoveryMetadata === 'string' ? cloudDiscoveryMetadata : JSON.stringify(cloudDiscoveryMetadata);
+            msalConfigWithMetadata.auth.authorityMetadata = typeof authorityMetadata === 'string' ? authorityMetadata : JSON.stringify(authorityMetadata);
+        } catch (error) {
+            console.log(error);
+        }
+
+        return msalConfigWithMetadata;
+    }
+
+    private static async fetchCloudDiscoveryMetadata(tenantId: string): Promise<any> {
```

Comment:
nit: Consider making non-static and reading the tenant id from config.

Reply:
the class is not instantiated yet at this point though :/